Mathlib/SetTheory/Zfc/Basic.lean

```diff
@@ -0,0 +1,1767 @@
+  ZFSet.sep (fun z => ∃ a ∈ x, ∃ b ∈ y, z = pair a b ∧ p a b) (powerset (powerset (x ∪ y)))
+#align Set.pair_sep ZFSet.pairSep
+
+@[simp]
+theorem mem_pairSep {p} {x y z : ZFSet.{u}} :
+    z ∈ pairSep p x y ↔ ∃ a ∈ x, ∃ b ∈ y, z = pair a b ∧ p a b :=
+  by
+  refine' mem_sep.trans ⟨And.right, fun e => ⟨_, e⟩⟩
+  rcases e with ⟨a, ax, b, bY, rfl, pab⟩
+  simp only [mem_powerset, subset_def, mem_union, pair, mem_pair]
+  rintro u (rfl | rfl) v <;> simp only [mem_singleton, mem_pair]
+  · rintro rfl
+    exact Or.inl ax
+  · rintro (rfl | rfl) <;> [left, right] <;> assumption
+#align Set.mem_pair_sep ZFSet.mem_pairSep
+
+theorem pair_injective : Function.Injective2 pair := fun x x' y y' H =>
+  by
+  have ae := ext_iff.1 H
+  simp only [pair, mem_pair] at ae
+  obtain rfl : x = x' := by
+    cases' (ae {x}).1 (by simp) with h h
+    · exact singleton_injective h
+    · have m : x' ∈ ({x} : ZFSet) := by simp [h]
+      rw [mem_singleton.mp m]
+  have he : x = y → y = y' := by
+    rintro rfl
+    cases' (ae {x, y'}).2 (by simp only [eq_self_iff_true, or_true_iff]) with xy'x xy'xx
+    · rw [eq_comm, ← mem_singleton, ← xy'x, mem_pair]
+      exact Or.inr rfl
+    · simpa [eq_comm] using (ext_iff.1 xy'xx y').1 (by simp)
+  obtain xyx | xyy' := (ae {x, y}).1 (by simp)
+  · obtain rfl := mem_singleton.mp ((ext_iff.1 xyx y).1 <| by simp)
+    simp [he rfl]
+  · obtain rfl | yy' := mem_pair.mp ((ext_iff.1 xyy' y).1 <| by simp)
+    · simp [he rfl]
+    · simp [yy']
+#align Set.pair_injective ZFSet.pair_injective
+
+@[simp]
+theorem pair_inj {x y x' y' : ZFSet} : pair x y = pair x' y' ↔ x = x' ∧ y = y' :=
+  pair_injective.eq_iff
+#align Set.pair_inj ZFSet.pair_inj
+
+/-- The cartesian product, `{(a, b) | a ∈ x, b ∈ y}` -/
+def prod : ZFSet.{u} → ZFSet.{u} → ZFSet.{u} :=
+  pairSep fun _ _ => True
+#align Set.prod ZFSet.prod
+
+@[simp]
+theorem mem_prod {x y z : ZFSet.{u}} : z ∈ prod x y ↔ ∃ a ∈ x, ∃ b ∈ y, z = pair a b := by
+  simp [prod]
+#align Set.mem_prod ZFSet.mem_prod
+
+theorem pair_mem_prod {x y a b : ZFSet.{u}} : pair a b ∈ prod x y ↔ a ∈ x ∧ b ∈ y :=
+  by
+  simp
+#align Set.pair_mem_prod ZFSet.pair_mem_prod
+
+/-- `isFunc x y f` is the assertion that `f` is a subset of `x × y` which relates to each element
+of `x` a unique element of `y`, so that we can consider `f` as a ZFC function `x → y`. -/
+def IsFunc (x y f : ZFSet.{u}) : Prop :=
+  f ⊆ prod x y ∧ ∀ z : ZFSet.{u}, z ∈ x → ∃! w, pair z w ∈ f
+#align Set.is_func ZFSet.IsFunc
+
+/-- `funs x y` is `y ^ x`, the set of all set functions `x → y` -/
+def funs (x y : ZFSet.{u}) : ZFSet.{u} :=
+  ZFSet.sep (IsFunc x y) (powerset (prod x y))
+#align Set.funs ZFSet.funs
+
+@[simp]
+theorem mem_funs {x y f : ZFSet.{u}} : f ∈ funs x y ↔ IsFunc x y f := by simp [funs, IsFunc]
+#align Set.mem_funs ZFSet.mem_funs
+
+/- Porting note: the `Definable` argument in `mapDefinableAux` is unused, though the TODO remark
+   suggests it shouldn't be. -/
+set_option linter.unusedVariables false
```

Comment:
This doesn't actually disable the linter. Am I using the wrong option?

Reply:
There are actually two unused variable linters. One of them reports immediately on affected variables and is controlled by ` set_option linter.unusedVariables` (this is the lean core linter), and the other one reports on `#lint` and is controlled by `@[nolint unusedVariables]` (this is the mathlib linter).